runtime/compiler/runtime/JITClientSession.cpp

```diff
@@ -375,6 +437,385 @@ ClientSessionData::cacheIProfilerInfo(TR_OpaqueMethodBlock *method, uint32_t byt
+   J9MethodInfo *methodInfo = NULL;
+   const AOTCacheMethodRecord *methodRecord = NULL;
+      {
+      OMR::CriticalSection cs(getROMMapMonitor());
+      auto it = getJ9MethodMap().find((J9Method *)method);
+      TR_ASSERT_FATAL(it != getJ9MethodMap().end(), "Method %p must be already cached", method);
+      methodInfo = &it->second;
+
+      // Obtain or create the methodRecord for this j9method. This assumes we have the ROMMapMonitor
+      methodRecord = getMethodRecord(*methodInfo, (J9Method *)method);
+      if (!methodRecord)
+         {
+         if (TR::Options::getVerboseOption(TR_VerboseJITServerSharedProfile))
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+               "ERROR: Failed to get method record for profiled method %p while compiling %s for clientUID %llu",
+               method, comp->signature(), (unsigned long long)_clientUID);
+         _numSharedProfileCacheMethodStoresFailed++;
+         return false; // serious error
+         }
+      }
+   TR::StackMemoryRegion region(*comp->trMemory());
+   Vector<TR_IPBytecodeHashTableEntry *> entries(region);
+   Vector<TR_IPBCDataCallGraph *>        cgEntries(region);
+   // aotCgEntries are duplicates of cgEntries where the j9class pointers are replaced with class records
+   Vector<TR_IPBCDataCallGraph *>        aotCgEntries(region);
+   // Go through the entire ipData that we received from the client in serialized format,
+   // deserialize the data and create entries ready to be added to the cache
+   uintptr_t methodStart = TR::Compiler->mtd.bytecodeStart(method);
+   auto bufferPtr = (uint8_t *)ipData.data();
+   while (bufferPtr < (uint8_t *)ipData.data() + ipData.size())
+      {
+      auto storage = (TR_IPBCDataStorageHeader *)bufferPtr;
+      // Note: the entries will contain the byteCodeIndex instead of the PC
+      auto entry = JITServerIProfiler::ipBytecodeHashTableEntryFactory(storage, methodStart+storage->pc, comp->trMemory(),
+                                                                       stackAlloc);
+      entry->deserialize(storage);
+      if (storage->ID == TR_IPBCD_CALL_GRAPH)
+         cgEntries.push_back((TR_IPBCDataCallGraph *)entry);
+      else
+         entries.push_back(entry);
+
+      if (!storage->left)
+         break;
+      bufferPtr += storage->left;
+      }
+
+   bool stackAllocatedEntries = true;
+
+   // If there are CallGraph entries, create duplicates of these entries, replace the
+   // J9class pointers in these entries with cache class records and add the new entries to aotCgEntries
+   // TODO: create a bigger critical section since we acquire ROMMapMonitor above too.
+   if (!cgEntries.empty())
+      {
+      // TODO: fill in this code when callGraph entries are passed in by the client
+      }
+   // Add the transformed callGraph entries to the branch/switch entries set.
+   entries.insert(entries.end(), aotCgEntries.begin(), aotCgEntries.end());
```

Comment:
Will this be done in some future PR? As of this PR, `aotCgEntries` is empty.

Reply:
Yes, this will be implemented in the next PR which will handle CallGraph entries.